File: capabilities/auth/targets/angular/files/src/app/features/auth/auth-api.ts
import { HttpClient, HttpContext, HttpContextToken } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { APP_BASE_URL } from '../../core/config/api-config';
import { AuthError, GENERIC_CREDENTIALS_MESSAGE, GENERIC_UNAVAILABLE_MESSAGE } from './auth-errors';
import type { AuthSession } from './auth-session';

/**
 * Marks a request as belonging to authentication itself.
 *
 * The auth interceptor reads it to leave these calls alone: retrying a failed
 * refresh through the very interceptor that triggered it is how a client ends up
 * in an infinite refresh loop.
 */
export const AUTH_ENDPOINT = new HttpContextToken<boolean>(() => false);

function authContext(): HttpContext {
  return new HttpContext().set(AUTH_ENDPOINT, true);
}

/** Talks to the authority. Holds no state and decides nothing about sessions. */
@Injectable({ providedIn: 'root' })
export class AuthApi {
  readonly #http = inject(HttpClient);
  readonly #baseUrl = inject(APP_BASE_URL);

  async login(email: string, password: string): Promise<AuthSession> {
    return this.#call('/api/v1/auth/login', { email, password }, GENERIC_CREDENTIALS_MESSAGE);
  }

  async refresh(refreshToken: string): Promise<AuthSession> {
    return this.#call('/api/v1/auth/refresh', { refreshToken }, GENERIC_UNAVAILABLE_MESSAGE);
  }

  /**
   * Best-effort remote revocation. The caller purges locally whatever happens:
   * an unreachable authority must never trap a user in a signed-in state.
   */
  async logout(refreshToken: string): Promise<void> {
    try {
      await firstValueFrom(this.#http.post(`${this.#baseUrl}/api/v1/auth/logout`,
        { refreshToken }, { context: authContext() }));
    } catch {
      // Deliberately swallowed — see the contract above.
    }
  }

  async #call(path: string, body: unknown, failureMessage: string): Promise<AuthSession> {
    try {
      return await firstValueFrom(this.#http.post<AuthSession>(
        `${this.#baseUrl}${path}`, body, { context: authContext() },
      ));
    } catch (error: unknown) {
      // The response body may carry the reason; it is not surfaced. Only the
      // correlation id crosses, so a support request stays traceable without
      // telling the caller which part of the credential was wrong.
      const requestId = extractRequestId(error);
      throw new AuthError(failureMessage, requestId);
    }
  }
}

function extractRequestId(error: unknown): string | null {
  const candidate = (error as { error?: { requestId?: unknown } })?.error?.requestId;
  return typeof candidate === 'string' ? candidate : null;
}

File: capabilities/auth/targets/flutter/files/lib/src/features/auth/auth_api.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/api/dio_client.dart';
import '../../core/api/dio_provider.dart';
import 'auth_errors.dart';
import 'auth_session.dart';

/// Talks to the authority. Holds no state and decides nothing about sessions.
///
/// Left implementable on purpose: the controller's guarantees are about what it
/// does when the transport misbehaves, and a test can only express that by
/// substituting one that does.
class AuthApi {
  const AuthApi(this._transport);

  final Dio _transport;

  Future<AuthSession> login(String email, String password) => _exchange(
    '/api/v1/auth/login',
    {'email': email, 'password': password},
    genericCredentialsMessage,
  );

  Future<AuthSession> refresh(String refreshToken) => _exchange(
    '/api/v1/auth/refresh',
    {'refreshToken': refreshToken},
    genericUnavailableMessage,
  );

  /// Best-effort remote revocation. The caller purges locally whatever happens:
  /// an unreachable authority must never trap a user in a signed-in state.
  Future<void> logout(String refreshToken) async {
    try {
      await _transport.post<Map<String, dynamic>>(
        '/api/v1/auth/logout',
        data: {'refreshToken': refreshToken},
      );
    } catch (_) {
      // Deliberately swallowed — see the contract above. The guarantee itself
      // lives in AuthController.signOut, not here, so that swapping this
      // transport cannot silently remove it.
    }
  }

  Future<AuthSession> _exchange(
    String path,
    Map<String, dynamic> body,
    String failureMessage,
  ) async {
    try {
      final response = await _transport.post<Map<String, dynamic>>(
        path,
        data: body,
      );
      return AuthSession.fromJson(response.data ?? const <String, dynamic>{});
    } catch (error) {
      // The response body may carry the reason; it is never surfaced. Only the
      // correlation id crosses.
      throw AuthError(failureMessage, requestId: _requestIdOf(error));
    }
  }
}

String? _requestIdOf(Object error) {
  if (error is! DioException) return null;
  final body = error.response?.data;
  if (body is! Map) return null;
  final requestId = body['requestId'];
  return requestId is String ? requestId : null;
}

/// Transport used by authentication itself, and by the single retry that follows
/// a refresh.
///
/// Deliberately a *separate* client carrying no authentication interceptor:
/// sending a refresh through the interceptor that triggered it is how a client
/// loops against its own authority. Separation makes that loop unrepresentable
/// rather than merely guarded against.
final authTransportProvider = Provider<Dio>(
  (ref) => createDioClient(config: ref.watch(apiConfigProvider)),
);

final authApiProvider = Provider<AuthApi>(
  (ref) => AuthApi(ref.watch(authTransportProvider)),
);
